packages/react/src/components/Table/TableColumnCustomizationModal/TableColumnCustomizationModal.jsx

```diff
@@ -0,0 +1,589 @@
+        ? transformToSelectedGroupItem(initialOrdering, groupItem)
+        : findAndCloneInitialTopLevelItem(columnId, availableColumnItems);
+
+      if (selectedItem?.hasLoadMore) delete selectedItem.hasLoadMore;
+
+      return selectedItem;
+    })
+    .filter((column) => (column?.children ? column?.children.length : column));
+
+  return orderedSelectedColumnItems.length
+    ? uniqBy(orderedSelectedColumnItems, (item) => item.id)
+    : [];
+};
+
+const preventDropInOtherGroup = (...args) => args[2] !== 'nested';
+
+const TableColumnCustomizationModal = ({
+  error,
+  groupMapping,
+  hasLoadMore,
+  hasVisibilityToggle,
+  i18n,
+  initialOrdering,
+  isPrimaryButtonDisabled,
+  availableColumns,
+  loadingMoreIds,
+  onChange,
+  onClearError,
+  onClose,
+  onLoadMore,
+  onReset,
+  onSave: onSaveCallback,
+  open,
+  overrides,
+  pinnedColumnId,
+  primaryValue,
+  secondaryValue,
+  showLoaderInAvailableList,
+  showLoaderInSelectedList,
+  testId,
+}) => {
+  const {
+    availableColumnsEmptyText,
+    availableColumnsLabel,
+    cancelButtonLabel,
+    clearSearchIconDescription,
+    closeIconDescription,
+    collapseIconDescription,
+    expandIconDescription,
+    loadMoreButtonLabel,
+    modalTitle,
+    modalBody,
+    removeIconDescription,
+    resetButtonLabel,
+    saveButtonLabel,
+    searchPlaceholder,
+    selectedColumnsEmptyText,
+    selectedColumnsLabel,
+    hideIconDescription,
+    showIconDescription,
+  } = merge({}, defaultProps.i18n, i18n);
+  const nrOfItemsNotNeedingSearch = 12;
+
+  const availableColumnItems = useMemo(
+    () =>
+      availableColumns.length
+        ? transformToAvailableItems({
+            availableColumns,
+            pinnedColumnId,
+            hasLoadMore,
+            groupMapping,
+            primaryValueKey: primaryValue,
+            secondaryValueKey: secondaryValue,
+          })
+        : [],
+    [availableColumns, pinnedColumnId, hasLoadMore, primaryValue, secondaryValue, groupMapping]
+  );
+
+  const [searchValue, setSearchValue] = useState(null);
+  const [hiddenIds, setHiddenIds] = useState(
+    initialOrdering.filter((col) => col.isHidden).map((col) => col.columnId)
+  );
+  const [selectedColumnItems, setSelectedColumnItems] = useState(() => {
+    return !availableColumnItems.length
+      ? []
+      : transformToSelectedItems(initialOrdering, availableColumnItems, groupMapping);
+  });
+
+  const initialSelectionForAsyncLoadedColumnsRequired = useRef(
+    initialOrdering.length && !availableColumnItems.length
+  );
+
+  useEffect(() => {
+    if (initialSelectionForAsyncLoadedColumnsRequired.current && availableColumnItems.length) {
+      initialSelectionForAsyncLoadedColumnsRequired.current = false;
+      setSelectedColumnItems(
+        transformToSelectedItems(initialOrdering, availableColumnItems, groupMapping)
+      );
+    }
+  }, [initialOrdering, availableColumnItems, groupMapping, selectedColumnItems]);
+
+  const onSave = () => {
```

Comment:
Should this also be wrapped in a `useCallback`

Reply:
I don't think so. There is no optimisation that rely on reference equality of that function in the child component (ComposedModal). It is simply forwarded to the onClick of the primary Button.